Source/Orts.Simulation/Simulation/RollingStocks/SubSystems/Brakes/MSTS/AirSinglePipe.cs

```diff
@@ -798,29 +1024,70 @@ public override void Update(float elapsedClockSeconds)
-                    dp = AutoCylPressurePSI - minCylPressurePSI;
+                if (UniformReleaseRatio > 0) // Uniform release: Brake release is slowed down when the brake pipe is substantially higher than the aux res
+                {
+                    if (!UniformReleaseActive && AuxResPressurePSI < BrakeLine1PressurePSI - UniformReleaseThresholdPSI)
+                        UniformReleaseActive = true;
+                    else if (UniformReleaseActive && AuxResPressurePSI > BrakeLine1PressurePSI - UniformReleaseThresholdPSI / 2)
+                        UniformReleaseActive = false;
+                    if (UniformReleaseActive)
+                        dp /= UniformReleaseRatio;
+                }
+                if (threshold > 0 && AutoCylPressurePSI < threshold + 1)
+                    dp *= MathHelper.Lerp(0.1f, 1.0f, AutoCylPressurePSI - threshold); // Reduce release rate if nearing target pressure to prevent toggling between release and lap
+                if (AutoCylPressurePSI - dp < threshold)
+                    dp = AutoCylPressurePSI - threshold;
                 if (dp < 0)
                     dp = 0;
                 AutoCylPressurePSI -= dp;
             }
-            
+            // Special cases for equipment which bypasses triple valve
+            else if (TwoStageLowPressureActive && AutoCylPressurePSI > TwoStageLowPressurePSI) // Two stage braking
+            {
+                float dp = elapsedClockSeconds * ReleaseRatePSIpS;
+                if (AutoCylPressurePSI - dp < TwoStageLowPressurePSI)
+                    dp = AutoCylPressurePSI - TwoStageLowPressurePSI;
+                AutoCylPressurePSI -= dp;
+            }
+            if (HighSpeedReducingPressurePSI > 0 && AutoCylPressurePSI > HighSpeedReducingPressurePSI) // High speed reducing valve
+            {
+                float dp = elapsedClockSeconds * MaxApplicationRatePSIpS / 2.0f; // This release rate should allow only emergency applications to overcome HSRV
+                dp *= MathHelper.Clamp(MathHelper.Lerp(1.0f, 0.1f, (AutoCylPressurePSI - HighSpeedReducingPressurePSI) / 5.0f), 0.1f, 1.0f); // Rate of release reduces as pressure difference increases
+                if (AutoCylPressurePSI - dp < HighSpeedReducingPressurePSI)
+                    dp = AutoCylPressurePSI - HighSpeedReducingPressurePSI;
+                AutoCylPressurePSI -= dp;
+            }
+
             // Manage emergency res charging
             if ((Car as MSTSWagon).EmergencyReservoirPresent)
             {
                 if (TripleValveState == ValveState.Release && EmergResPressurePSI > BrakeLine1PressurePSI)
                 {
-                    if (EmergResQuickRelease && AutoCylPressurePSI > 5) // Quick release: Emergency res charges brake pipe during release
+                    if (EmergResQuickRelease && QuickReleaseActive) // Quick release: Emergency res charges brake pipe during release
```

Comment:
UIC distributors have another feature that is also called quick release. I'd suggest renaming QuickReleaseActive to EmergQuickRelease active to avoid confusion.

UIC quick release is not implemented yet, IIRC the idea is that BP is overcharged to trigger a valve that fully releases all brakes, and then BP pressure quickly returns to normal